src/main/java/com/klodnicki/bike/rest/API/Bike/service/BikeService.java

```diff
@@ -28,18 +39,86 @@ public Bike findBikeById(Long id) throws NotFoundInDatabaseException {
+        if (userToBeAssigned == null) {
+            throw new NotFoundInDatabaseException();
+        }
+
+        //check if provided station is null, if yes throw an exception
+        ChargingStation chargingStationToBeAssigned = bikeToUpdate.getChargingStation();
+        if (chargingStationToBeAssigned == null) {
+            throw new NotFoundInDatabaseException();
+        }
+
+        // check if provided user exists. if not throw an exception if yes follow the logic:
+        // set user to the bike -> setting on the both sides because relation @OneToOne with owning on one side
+        if (userService.checkIfUserExistInDatabase(userToBeAssigned.getId())) {
+            bike.setUser(userToBeAssigned);
+            bikeToUpdate.getUser().setBike(bike);
+        }
+
+//     logic if bike is rented->removing bike from station, free slots of the station+1,
+//      make charging station field at this bike null
+        if (bikeToUpdate.isRented()) {
+            if (bike.getChargingStation() != null) {
+                bike.getChargingStation().getBikeList().remove(bikeToUpdate);
+                int freeSlots = bike.getChargingStation().getFreeSlots();
+                bike.getChargingStation().setFreeSlots(freeSlots + 1);
+            }
+            bike.setChargingStation(null);
+
+            //logic if bike is NOT rented -> checking if provided station exists, if no -> throws exception if yes logic:
+            // 1. set the charging station to the bike
+            // 2. merge the detached instance back into the current session by using the merge() method provided by the EntityManager
+            // 3. adds @Transactional to the method. This annotation will start a new transaction before the method is executed
+            // and commit it after the method returns.
+            // 4. remove user from the bike, user field at the bike should be null
+        } else if (chargingStationService.checkIfChargingStationExistInDatabase(chargingStationToBeAssigned.getId())) {
+            {
+                ChargingStation managedChargingStation = entityManager.merge(bikeToUpdate.getChargingStation());
+                bike.setChargingStation(managedChargingStation);
 
+                // Get the User object associated with this Bike
+                User user = bike.getUser();
+
+                // Set both sides of the relationship to null (because it is @OneToOne relation with owning one-sided)
+                bike.setUser(null); // one side
+                if (user != null) {
+                    user.setBike(null); // second side
+                    userService.addUser(user);  // Save the User to persist changes
+                }
+            }
+        }
         return bikeRepository.save(bike);
     }
+
+    public BikeForNormalUserDTO updateBikeUser(Long id, Bike bikeToUpdate) throws NotFoundInDatabaseException {
+        Bike bike = bikeRepository.findById(id).orElseThrow(NotFoundInDatabaseException::new);
```

Comment:
`Bike bike = bikeRepository.findById(id).orElseThrow(NotFoundInDatabaseException::new);`
Code repetition. There is already a method which does the same.

Reply:
OK! I replace it! thanks for that!